File: backend/app/schemas/knowledge.py
"""Knowledge schemas - 知识点与图谱"""
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import BaseResponse


class KnowledgePointBase(BaseModel):
    """知识点基础信息 (教研数据, 只读)"""
    code: str = Field(description="唯一 code, e.g. MATH-G3-CH1-K001")
    subject: str
    grade: int = Field(ge=1, le=6)
    chapter: str
    name: str
    description: Optional[str] = None

    prerequisites: list[str] = Field(default_factory=list, description="前置知识点 code")
    successors: list[str] = Field(default_factory=list, description="后续知识点 code")
    related: list[str] = Field(default_factory=list, description="相关知识点 code")

    importance: int = Field(default=3, ge=1, le=5)
    difficulty: int = Field(default=3, ge=1, le=5)

    video_urls: list[str] = Field(default_factory=list, description="关联微课 URL (basic.sh.smartedu.cn)")
    common_errors: list[str] = Field(default_factory=list, description="常见错误模式")


class KnowledgePointResponse(KnowledgePointBase, BaseResponse):
    """知识点详情"""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    status: str  # draft / published / deprecated
    created_at: datetime
    updated_at: datetime


class KnowledgePointListItem(BaseModel):
    """知识点列表项"""
    model_config = ConfigDict(from_attributes=True)

    code: str
    subject: str
    grade: int
    chapter: str
    name: str
    importance: int
    difficulty: int


class GraphNode(BaseModel):
    """图谱节点 (含孩子掌握度)"""
    code: str
    name: str
    chapter: str
    importance: int
    difficulty: int
    # 孩子当前状态
    mastery: float = Field(default=0.0, ge=0.0, le=1.0, description="0-1, 0=未学, 1=完全掌握")
    error_count: int = Field(default=0, description="累计错题数")
    has_video: bool = False


class GraphEdge(BaseModel):
    """图谱边"""
    source: str = Field(description="源节点 code")
    target: str = Field(description="目标节点 code")
    relation: str = Field(description="prerequisite / successor / related")


class KnowledgeGraphResponse(BaseModel):
    """知识图谱响应 (用于力导向/树状可视化)"""
    subject: str
    grade: int
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    summary: dict = Field(
        default_factory=dict,
        description="汇总: total/mastered/in_progress/unstudied 节点数",
    )


class ChapterInfo(BaseModel):
    """章节信息"""
    chapter: str
    knowledge_points: list[KnowledgePointListItem]
    total_points: int


__all__ = [
    "KnowledgePointBase",
    "KnowledgePointResponse",
    "KnowledgePointListItem",
    "GraphNode",
    "GraphEdge",
    "KnowledgeGraphResponse",
    "ChapterInfo",
]

File: web/src/lib/knowledge.ts
/**
 * 知识图谱 API
 */
import { api } from './api';
import type { SuccessResponse } from './types';

export interface KnowledgePoint {
  code: string;
  subject: string;
  grade: number;
  chapter: string;
  name: string;
  importance: number;
  difficulty: number;
}

export interface KnowledgePointDetail extends KnowledgePoint {
  id: string;
  description: string | null;
  prerequisites: string[];
  successors: string[];
  related: string[];
  video_urls: string[];
  common_errors: string[];
  status: string;
  created_at: string;
  updated_at: string;
}

export interface GraphNode {
  code: string;
  name: string;
  chapter: string;
  importance: number;
  difficulty: number;
  mastery: number;
  error_count: number;
  has_video: boolean;
}

export interface GraphEdge {
  source: string;
  target: string;
  relation: 'prerequisite' | 'successor' | 'related';
}

export interface KnowledgeGraphData {
  subject: string;
  grade: number;
  nodes: GraphNode[];
  edges: GraphEdge[];
  summary: {
    total: number;
    mastered: number;
    in_progress: number;
    unstudied: number;
    weak: number;
  };
}

export interface ChapterInfo {
  chapter: string;
  knowledge_points: KnowledgePoint[];
  total_points: number;
}

export const knowledgeApi = {
  async listPoints(opts: { subject?: string; grade?: number; chapter?: string } = {}): Promise<KnowledgePoint[]> {
    const resp = await api.get<SuccessResponse<KnowledgePoint[]>>('/knowledge/points', {
      params: opts,
    });
    return resp.data.data;
  },

  async getPoint(code: string): Promise<KnowledgePointDetail> {
    const resp = await api.get<SuccessResponse<KnowledgePointDetail>>(`/knowledge/points/${code}`);
    return resp.data.data;
  },

  async graph(opts: { subject: string; grade: number; childId?: string }): Promise<KnowledgeGraphData> {
    const resp = await api.get<SuccessResponse<KnowledgeGraphData>>('/knowledge/graph', {
      params: {
        subject: opts.subject,
        grade: opts.grade,
        child_id: opts.childId,
      },
    });
    return resp.data.data;
  },

  async chapters(opts: { subject: string; grade: number }): Promise<ChapterInfo[]> {
    const resp = await api.get<SuccessResponse<ChapterInfo[]>>('/knowledge/chapters', {
      params: opts,
    });
    return resp.data.data;
  },
};
